pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java

```diff
@@ -0,0 +1,844 @@
+            _currentGroupId++;
+          }
+          return _currentGroupId < _globalGroupIdUpperBound;
+        }
+
+        @Override
+        public GroupKey next() {
+          if (!hasNext()) {
+            throw new NoSuchElementException();
+          }
+          _groupKey._groupId = _currentGroupId;
+          _groupKey._stringKey = getGroupKey(_currentGroupId);
+          _currentGroupId++;
+          return _groupKey;
+        }
+
+        @Override
+        public void remove() {
+          throw new UnsupportedOperationException();
+        }
+      };
+    }
+  }
+
+  private class IntMapBasedHolder implements RawKeyHolder {
+    private final Int2IntOpenHashMap _rawKeyToGroupIdMap = new Int2IntOpenHashMap();
+
+    public IntMapBasedHolder() {
+      _rawKeyToGroupIdMap.defaultReturnValue(INVALID_ID);
+    }
+
+    @Override
+    public void processSingleValue(int numDocs, @Nonnull int[] outGroupIds) {
+      for (int i = 0; i < numDocs; i++) {
+        int rawKey = 0;
+        for (int j = _numGroupByColumns - 1; j >= 0; j--) {
+          rawKey = rawKey * _cardinalities[j] + _singleValueDictIds[j][i];
+        }
+        outGroupIds[i] = getGroupId(rawKey);
+      }
+    }
+
+    @Nonnull
+    @Override
+    public int[] processMultiValue(int index) {
+      int[] rawKeys = getIntRawKeys(index);
+      int length = rawKeys.length;
+      int[] groupIds = new int[length];
+      for (int i = 0; i < length; i++) {
+        groupIds[i] = getGroupId(rawKeys[i]);
+      }
+      return groupIds;
+    }
+
+    private int getGroupId(int rawKey) {
+      int groupId = _rawKeyToGroupIdMap.get(rawKey);
+      if (groupId == INVALID_ID) {
+        groupId = _rawKeyToGroupIdMap.size();
+        _rawKeyToGroupIdMap.put(rawKey, groupId);
+      }
+      return groupId;
+    }
+
+    @Override
+    public int getGroupIdUpperBound() {
+      return _rawKeyToGroupIdMap.size();
+    }
+
+    @Nonnull
+    @Override
+    public Iterator<GroupKey> iterator() {
+      return new Iterator<GroupKey>() {
+        private final ObjectIterator<Int2IntMap.Entry> _iterator = _rawKeyToGroupIdMap.int2IntEntrySet().fastIterator();
+        private final GroupKey _groupKey = new GroupKey();
+
+        @Override
+        public boolean hasNext() {
+          return _iterator.hasNext();
+        }
+
+        @Override
+        public GroupKey next() {
+          Int2IntMap.Entry entry = _iterator.next();
+          _groupKey._groupId = entry.getIntValue();
+          _groupKey._stringKey = getGroupKey(entry.getIntKey());
+          return _groupKey;
+        }
+
+        @Override
+        public void remove() {
+          throw new UnsupportedOperationException();
+        }
+      };
+    }
+  }
+
+  /**
+   * Helper method to calculate raw keys that can fit into integer for the given index.
+   *
+   * @param index Index in block
+   * @return Array of integer raw keys
+   */
+  @SuppressWarnings("Duplicates")
```

Comment:
Nit: Any way to get rid of the duplicate? Or is it IDE's limitation for not being smart enough to realize that data types changed?